applications/nrf_desktop/src/modules/dvfs.c

```diff
@@ -135,38 +144,94 @@ static void handle_dvfs_error(int32_t err)
 	} else if (err == -EAGAIN) {
 		LOG_DBG("DVFS not initialized, trying again.");
 		timeout = DVFS_RETRY_INITIALIZATION_TIMEOUT;
+	} else if (err == -ETIMEDOUT) {
+		LOG_DBG("DVFS frequency change timed out, trying again.");
+		timeout = DVFS_RETRY_BUSY_TIMEOUT;
 	} else {
 		LOG_ERR("DVFS freq change returned with error: %d", err);
-		module_set_state(MODULE_STATE_ERROR);
-		module_state = STATE_ERROR;
-		cancel_dvfs_retry_work();
+		dvfs_fatal_error();
 		return;
 	}
 	(void) k_work_reschedule(&dvfs_retry.retry_work, timeout);
 }
 
-static void set_dvfs_freq(enum dvfs_frequency_setting target_freq)
+static uint32_t check_required_frequency(void)
+{
+	if (dfvs_requests_state_bitmask & dvfs_high_freq_bitmask) {
+		return CONFIG_DESKTOP_DVFS_FREQ_HIGH;
+	} else if (dfvs_requests_state_bitmask & dvfs_medlow_freq_bitmask) {
+		return CONFIG_DESKTOP_DVFS_FREQ_MED;
+	} else {
+		return CONFIG_DESKTOP_DVFS_FREQ_LOW;
+	}
+}
+
+static void dvfs_notify_cb(void)
+{
+	(void) k_work_submit(&dvfs_notify_work);
+}
+
+static void set_dvfs_freq(uint32_t target_freq)
 {
-	int32_t ret = dvfs_service_handler_change_freq_setting(target_freq);
+	int ret;
+
+	if (spec.frequency != 0) {
+		ret = nrf_clock_control_release(dvfs_clock_dev, &spec);
+		if (ret < 0) {
+			LOG_ERR("Failed to release requested clock specs, error: %d", ret);
+			dvfs_fatal_error();
+			return;
+		}
+	}
 
+	sys_notify_init_callback(&cli.notify, dvfs_notify_cb);
+
+	spec.frequency = target_freq;
+	ret = nrf_clock_control_request(dvfs_clock_dev, &spec, &cli);
 	if (ret) {
 		handle_dvfs_error(ret);
 	} else {
-		current_freq = target_freq;
-		LOG_INF("Have requested %s frequency",
-			get_dvfs_frequency_setting_name(target_freq));
-		cancel_dvfs_retry_work();
+		LOG_INF("Have requested %" PRIu32 " frequency", target_freq);
+		requested_freq = target_freq;
+		request_in_progress = true;
 	}
 }
 
-static enum dvfs_frequency_setting check_required_frequency(void)
+static void dvfs_notify_work_handler(struct k_work *work)
 {
-	if (dfvs_requests_state_bitmask & dvfs_high_freq_bitmask) {
-		return DVFS_FREQ_HIGH;
-	} else if (dfvs_requests_state_bitmask & dvfs_medlow_freq_bitmask) {
-		return DVFS_FREQ_MEDLOW;
-	} else {
-		return DVFS_FREQ_LOW;
+	int res;
+	int ret = sys_notify_fetch_result(&cli.notify, &res);
+
+	if (ret < 0) {
+		LOG_WRN("Work not completed, verify usage of notify API");
+		return;
+	}
+
+	request_in_progress = false;
+
+	if (res < 0) {
```

Comment:
don't understand you on both comments.

Reply:
I would like to avoid an early return here and use `if ... else` instead. Splitting out a small function for code that currently is placed after the early return would improve readability further.

Reply:
IMO we should return. If there is a problem we don't want to preform all other things. 
If you would put under if ... else.. we would have to place rest of code in else clause which I find more confusing then early return on encountering some problem.

Reply:
Ok, let's keep it as is